[docs] Improve configuration details of functions worker (#7534)
apache/pulsar: site2/docs/functions-runtime.md

	at io.kubernetes.client.ApiClient.execute(ApiClient.java:798) ~[io.kubernetes-client-java-2.0.0.jar:?]
	at io.kubernetes.client.apis.CoreV1Api.readNamespacedConfigMapWithHttpInfo(CoreV1Api.java:23673) ~[io.kubernetes-client-java-api-2.0.0.jar:?]
	at io.kubernetes.client.apis.CoreV1Api.readNamespacedConfigMap(CoreV1Api.java:23655) ~[io.kubernetes-client-java-api-2.0.0.jar:?]
	at org.apache.pulsar.functions.runtime.KubernetesRuntimeFactory.fetchConfigMap(KubernetesRuntimeFactory.java:284) [org.apache.pulsar-pulsar-functions-runtime-2.4.0-42c3bf949.jar:2.4.0-42c3bf949]
	at org.apache.pulsar.functions.runtime.KubernetesRuntimeFactory$1.run(KubernetesRuntimeFactory.java:275) [org.apache.pulsar-pulsar-functions-runtime-2.4.0-42c3bf949.jar:2.4.0-42c3bf949]
	at java.util.TimerThread.mainLoop(Timer.java:555) [?:1.8.0_212]
	at java.util.TimerThread.run(Timer.java:505) [?:1.8.0_212]
```

### Kubernetes Secrets integration

In order to safely distribute secrets, Pulasr Functions can reference Kubernetes secrets. To enable this, set the `secretsProviderConfiguratorClassName` to `org.apache.pulsar.functions.secretsproviderconfigurator.KubernetesSecretsProviderConfigurator`.

> Then, you can create a secret in the namespace where your functions will be deployed. As an example, suppose we are deploying functions to the `pulsar-func` Kubernetes namespace, and we have a secret named `database-creds` with a field name `password`, which we want to mount in the pod as an environment variable called `DATABASE_PASSWORD`.

Reviewer: ```suggestion
Then, you can create a secret in the namespace where your functions are deployed. As an example, suppose we are deploying functions to the `pulsar-func` Kubernetes namespace, and we have a secret named `database-creds` with a field name `password`, which we want to mount in the pod as an environment variable called `DATABASE_PASSWORD`.
```

The following functions configuration would then allow us to reference that secret and have the value be mounted as an environment variable in the pod.

```Yaml
tenant: "mytenant"
namespace: "mynamespace"
name: "myfunction"
topicName: "persistent://mytenant/mynamespace/myfuncinput"
className: "com.company.pulsar.myfunction"

secrets:
  # the secret will be mounted from the `password` field in the `database-creds` secret as an env var called `DATABASE_PASSWORD`
  DATABASE_PASSWORD:
    path: "database-creds"
    key: "password"

```

### Kubernetes Functions authentication

> When your Pulsar cluster uses authentication, the pod running your function needs a mechanism to authenticate with the Broker.

Reviewer: ```suggestion
When your Pulsar cluster uses authentication, the pod running your function needs a mechanism to authenticate with the broker.
```

An interface, `org.apache.pulsar.functions.auth.KubernetesFunctionAuthProvider`, can be extended to provide support for any authentication mechanism. The `functionAuthProviderClassName` in `function-worker.yml` is used to specify your path to this implementation.

Pulsar includes an implementation of this interface that is suitable for token auth, which should be configured like the following in the configuration:
```Yaml
functionAuthProviderClassName: org.apache.pulsar.functions.auth.KubernetesSecretsTokenAuthProvider
```

> For custom auth or TLS, the user will need to implement this interface (or use an alternative mechanism to provide auth)

Reviewer: ```suggestion
For custom authentication or TLS, you need to implement this interface (or use an alternative mechanism to provide authentication)
```

For token authentication, the way this works is that the functions worker captures the token that was used to deploy (or update) the function. This token is then saved as a secret and mounted into the pod.

One thing to keep in mind is that if you use tokens that expire when deploying functions, these tokens will expire.

### Kubernetes CustomRuntimeOptions

> The Kubernetes integration also includes the ability for the user to implement a class which will customize how manifests will be generated. This is configured by setting the `runtimeCustomizerClassName` in the `functions-worker.yml` to the fully qualified class name. The interface the user must implement is the `org.apache.pulsar.functions.runtime.kubernetes.KubernetesManifestCustomizer` interface.

Reviewer: ```suggestion
The Kubernetes integration also includes the ability for the user to implement a class which customizes how manifests is generated. This is configured by setting the `runtimeCustomizerClassName` in the `functions-worker.yml` to the fully qualified class name. The interface the you must implement is the `org.apache.pulsar.functions.runtime.kubernetes.KubernetesManifestCustomizer` interface.
```

IBM Writing Style Guide: Use the second-person pronoun `you` as much as possible. The subject of an imperative sentence is understood to be you.

> The functions (and sinks/sources) API provides a flag, `customRuntimeOptions` which will be passed to this interface.

Reviewer: ```suggestion
The functions (and sinks/sources) API provides a flag, `customRuntimeOptions`, which is passed to this interface.
```

  }
}
```

## Other configuration considerations

### In a cluster with geo-replication

> If you are running multiple clusters tied together with geo-replication, it is important to use a different function namespace for each cluster! Otherwise, the functions

Reviewer: ```suggestion
If you are running multiple clusters tied together with geo-replication, it is important to use a different function namespace for each cluster. Otherwise, the function
```

IBM Writing Style Guide: 
Do not punctuate sentences with exclamation points because their tone can be interpreted negatively,
for example, as aggressive, condescending, or overly informal.
Convey urgency or emphasis with the appropriate words, not with exclamation points. To call
attention to important hints, tips, guidance, restrictions, or advice that might be overlooked, consider
using a note that has a meaningful label.

> will share a namespace and potentially schedule across clusters.

Reviewer: ```suggestion
shares a namespace and potentially schedule across clusters.
```

As an example, suppose we have clusters east-1 and west-1, we would configure the Functions Worker in east-1 like:
```Yaml
pulsarFunctionsCluster: east-1
pulsarFunctionsNamespace: public/functions-east-1
```

and the cluster in west-1 like:
```Yaml
pulsarFunctionsCluster: west-1
pulsarFunctionsNamespace: public/functions-west-1
```

> This ensures the two different Function Workers will use distinct sets of topics for their internal co-ordination.

Reviewer: ```suggestion
This ensures the two different Function Workers use distinct sets of topics for their internal coordination.
```

> ### Configuring Standalone Functions Worker

Reviewer: ```suggestion
### Configure standalone Functions Worker
```

When configuring a standalone Functions Worker, you need to specify a few properties in order for the Functions Worker to be able

> to communicate with the Broker. This requires many of the same properties to be set that the Broker requires, especially when using TLS.

Reviewer: ```suggestion
to communicate with the broker. This requires many of the same properties to be set that the broker requires, especially when using TLS.
```

The following properties are the baseline of what is required:

```Yaml
workerPort: 8080
workerPortTls: 8443 # when using TLS
tlsCertificateFilePath: /etc/pulsar/tls/tls.crt # when using TLS
tlsKeyFilePath: /etc/pulsar/tls/tls.key # when using TLS
tlsTrustCertsFilePath: /etc/pulsar/tls/ca.crt # when using TLS
pulsarServiceUrl: pulsar://broker.pulsar:6650/ # or pulsar+ssl://pulsar-prod-broker.pulsar:6651/ when using TLS
pulsarWebServiceUrl: http://broker.pulsar:8080/ # or https://pulsar-prod-broker.pulsar:8443/ when using TLS
useTls: true # when using TLS, critical!

```

> #### With Authentication

Reviewer: ```suggestion
#### With authentication
```

> When running a Functions Worker in a standalone process (i.e. not embedded in the Broker) in a cluster with authentication, you must configure your Functions Worker to both be able to interact with the Broker *and* also to authenticate incoming requests as well. This requires many of the same properties to be set that the Broker requires for authentication/authorization.

Reviewer: ```suggestion
When running a Functions Worker in a standalone process (that is, not embedded in the broker) in a cluster with authentication, you must configure your Functions Worker to both be able to interact with the broker *and* also to authenticate incoming requests as well. This requires many of the same properties to be set that the broker requires for authentication or authorization.
```

> As an example, assuming we want to use token authentication, here is an example of properties we need to set in `function-worker.yml`

Reviewer: ```suggestion
As an example, assuming you want to use token authentication. Here is an example of properties you need to set in `function-worker.yml`
```

IBM Writing Style Guide:
• Avoid the first-person pronouns `I` and `we`, except in these situations:
° In the question portion of frequently asked questions (FAQs)
° In articles, white papers, or documents that have listed authors and in which the authors describe their own actions or opinions

```Yaml
clientAuthenticationPlugin: org.apache.pulsar.client.impl.auth.AuthenticationToken
clientAuthenticationParameters: file:///etc/pulsar/token/admin-token.txt
configurationStoreServers: zookeeper-cluster:2181 # auth requires a connection to zookeeper
authenticationProviders:
 - "org.apache.pulsar.broker.authentication.AuthenticationProviderToken"
authorizationEnabled: true
authenticationEnabled: true
superUserRoles:
  - superuser
  - proxy
properties:
  tokenSecretKey: file:///etc/pulsar/jwt/secret # if using a secret token
  tokenPublicKey: file:///etc/pulsar/jwt/public.key # if using public/private key tokens
```

> Notice that we must configure both the Function Worker authz/authn for the server to proper auth requests and configure the client to be authenticated to communicate with the broker.

Reviewer: ```suggestion
> #### Note 
>
> You must configure both the Function Worker authorization or authentication for the server to proper authentication requests and configure the client to be authenticated to communicate with the broker.
```
